crates/turborepo-lib/src/package_graph/mod.rs

```diff
@@ -236,6 +239,71 @@ impl PackageGraph {
             .flatten()
             .collect()
     }
+
+    /// Returns a list of changed packages based on the contents of a previous
+    /// `Lockfile`. This assumes that none of the package.json in the workspace
+    /// change, it is the responsibility of the caller to verify this.
+    pub fn changed_packages(
+        &self,
+        previous: &dyn Lockfile,
+    ) -> Result<Vec<WorkspaceName>, ChangedPackagesError> {
+        let current = if let Some(lockfile) = self.lockfile() {
+            lockfile
+        } else {
+            return Err(ChangedPackagesError::NoLockfile);
+        };
+
+        let external_deps = self
+            .workspaces()
+            .filter_map(|(a, b)| {
+                b.unresolved_external_dependencies.as_ref().map(|b| {
+                    (
+                        a.to_string(),
+                        b.iter()
+                            .map(|pkg| (pkg.name.to_owned(), pkg.version.to_owned()))
+                            .collect(),
+                    )
+                })
+            })
+            .collect::<HashMap<_, HashMap<_, _>>>();
+
+        let closures = turborepo_lockfiles::all_transitive_closures(previous, external_deps)?;
+
+        let current_key = current.global_change_key();
+        let previous_key = previous.global_change_key();
+
+        let global_change = current_key != previous_key;
+
+        let (changed, root_change) = self
+            .workspaces
+            .iter()
+            .filter(|(name, info)| {
+                closures.get(info.package_json_path().as_str())
+                    != info.transitive_dependencies.as_ref()
+            })
+            .fold((vec![], false), |(mut changed, all), (name, info)| {
+                if WorkspaceName::Root.eq(name) {
```

Comment:
Total bikeshed, I prefer `match`/`matches!` compared to using `==` when we are checking against a specific variant
```suggestion
                if matches!(name, WorkspaceName::Root) {
```